web/frontend/src/components/slam.vue

```diff
@@ -1,193 +1,221 @@
 
 <script setup lang="ts">
 
-// import { nextTick } from 'vue';
+import { $ref, $computed } from 'vue/macros';
 import { grpc } from '@improbable-eng/grpc-web';
-import { Client, commonApi, slamApi } from '@viamrobotics/sdk';
-import { displayError } from '../lib/error';
+import { Client, commonApi, ServiceError, slamApi } from '@viamrobotics/sdk';
+import { displayError, isServiceError } from '../lib/error';
 import { rcLogConditionally } from '../lib/log';
+import PCD from './pcd/pcd-view.vue';
+import Slam2dRender from './slam-2d-render.vue';
 
-/*
- * import PCD from './pcd/pcd-view.vue';
- * import Slam2dRender from './slam-2d-render.vue';
- */
+type MapAndPose = { map: Uint8Array, pose: commonApi.Pose}
 
 const props = defineProps<{
   name: string
   resources: commonApi.ResourceName.AsObject[]
   client: Client
 }>();
 
-let imageMap = $ref('');
 const selected2dValue = $ref('manual');
+const selected3dValue = $ref('manual');
+let pointCloudUpdateCount = $ref(0);
+let pointcloud = $ref<Uint8Array | undefined>();
+let pose = $ref<commonApi.Pose | undefined>();
 let show2d = $ref(false);
-let slam2dIntervalId = -1;
+let show3d = $ref(false);
+let refresh2DCancelled = true;
+let refresh3DCancelled = true;
 
-/*
- * const selected3dValue = $ref('manual');
- * let pointCloudUpdateCount = $ref(0);
- * let pointcloud = $ref<Uint8Array | undefined>();
- * let pose = $ref<commonApi.Pose | undefined>();
- * let show3d = $ref(false);
- * const loaded2d = $computed(() => (pointcloud !== undefined && pose !== undefined));
- * let slam3dIntervalId = -1;
- */
+const loaded2d = $computed(() => (pointcloud !== undefined && pose !== undefined));
 
-/*
- * const concatArrayU8 = (arrays: Uint8Array[]) => {
- *   const totalLength = arrays.reduce((acc, value) => acc + value.length, 0);
- *   const result = new Uint8Array(totalLength);
- *   let length = 0;
- *   for (const array of arrays) {
- *     result.set(array, length);
- *     length += array.length;
- *   }
- *   return result;
- * };
- */
+let slam2dTimeoutId = -1;
+let slam3dTimeoutId = -1;
 
-const fetchSLAMImageMap = (name: string) => {
-  const req = new slamApi.GetMapRequest();
-  req.setName(name);
-  req.setMimeType('image/jpeg');
-  req.setIncludeRobotMarker(true);
-  rcLogConditionally(req);
-  props.client.slamService.getMap(req, new grpc.Metadata(), (error, response) => {
-    if (error) {
-      return displayError(error);
-    }
-    const blob = new Blob([response!.getImage_asU8()], { type: 'image/jpeg' });
-    imageMap = URL.createObjectURL(blob);
-  });
+const concatArrayU8 = (arrays: Uint8Array[]) => {
```

Comment:
[q] Is there any reason `flat` can't work here (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat)